libeos-update-server/server.c

```diff
@@ -0,0 +1,401 @@
+                         gint       delta)
+{
+  GObject *obj = G_OBJECT (self);
+
+  g_assert (delta >= 0 || self->pending_requests >= (guint) -delta);
+  g_assert (delta <= 0 || self->pending_requests <= G_MAXUINT - delta);
+
+  if (delta == 0)
+    return;
+
+  g_debug ("%s: Updating from %u to %u", G_STRFUNC, self->pending_requests,
+           self->pending_requests + delta);
+
+  self->pending_requests += delta;
+  self->last_request_time = g_get_monotonic_time ();
+
+  g_object_freeze_notify (obj);
+  g_object_notify_by_pspec (obj, props[PROP_PENDING_REQUESTS]);
+  g_object_notify_by_pspec (obj, props[PROP_LAST_REQUEST_TIME]);
+  g_object_thaw_notify (obj);
+}
+
+static void
+request_read_cb (SoupServer        *soup_server,
+                 SoupMessage       *message,
+                 SoupClientContext *client,
+                 gpointer           user_data)
+{
+  EusServer *self = EUS_SERVER (user_data);
+
+  update_pending_requests (self, 1);
+}
+
+static void
+request_finished_cb (SoupServer        *soup_server,
+                     SoupMessage       *message,
+                     SoupClientContext *client,
+                     gpointer           user_data)
+{
+  EusServer *self = EUS_SERVER (user_data);
+
+  update_pending_requests (self, -1);
+}
+
+static void
+request_aborted_cb (SoupServer        *soup_server,
+                    SoupMessage       *message,
+                    SoupClientContext *client,
+                    gpointer           user_data)
+{
+  EusServer *self = EUS_SERVER (user_data);
+
+  update_pending_requests (self, -1);
+}
+
+/**
+ * eus_server_new:
+ * @server: #SoupServer to handle requests from
+ *
+ * Create a new #EusServer to handle requests from @server.
+ *
+ * Returns: (transfer full): The server.
+ */
+EusServer *
+eus_server_new (SoupServer *server)
+{
+  g_return_val_if_fail (SOUP_IS_SERVER (server), NULL);
+
+  return g_object_new (EUS_TYPE_SERVER,
+                       "server", server,
+                       NULL);
+}
+
+/**
+ * eus_server_add_repo:
+ * @self: an #EusServer
+ * @repo: repository to start serving
+ *
+ * Add an #EusRepo to the server, and immediately make its contents available
+ * to clients of the server.
+ *
+ * The repository will be available until eus_server_disconnect() is called.
+ *
+ * Since: UNRELEASED
+ */
+void
+eus_server_add_repo (EusServer *self,
+                     EusRepo   *repo)
+{
+  g_return_if_fail (EUS_IS_SERVER (self));
+  g_return_if_fail (EUS_IS_REPO (repo));
+
+  g_ptr_array_add (self->repos, g_object_ref (repo));
+  eus_repo_connect (repo, self->server);
+}
+
+/**
+ * eus_server_disconnect:
+ * @self: an #EusServer
+ *
+ * Disconnect the server and all its repositories from the underlying
+ * #SoupServer and its socket. Cancel all pending requests and stop handling
+ * any new ones.
+ *
+ * This does not call soup_server_disconnect() on the underlying #SoupServer.
+ *
+ * This is called automatically when the #EusServer is disposed.
```

Comment:
This is actually a lie. I think that the call to the `eus_server_disconnect` function should happen in dispose before `GPtrArray` holding the repos is unreffed.